### PR TITLE
Add various webcomponents 1.0 externs

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
Adds externs for Custom Elements as well as some other misc related definitions.

Initially based off https://github.com/Polymer/polymer/blob/3eb62e8fe9256f8f0093a7d53c1f8c4dc14888ea/externs/closure-upstream-externs.js

cc @justinfagnani

### CODE REVIEW

File: externs/browser/html5.js [hunk 1]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
 
 /** @type {string} */
 Plugin.prototype.name;
+
+/** @record */
+function CustomElementInterface() {}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Why both this and `CustomElement`?

File: externs/browser/html5.js [hunk 2]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+CustomElement.prototype.disconnectedCallback;
+
+/** @type {function(string, ?string, ?string, ?string)|undefined} */
+CustomElement.prototype.attributeChangedCallback;
+
+/**
+ * @see https://w3c.github.io/webcomponents/spec/custom/#custom-elements-api
+ * @const
+ */
+var customElements = {};
```
Comment (justinfagnani) [inline externs/browser/html5.js]: I think this should be of type `CustomElementRegistry`, which is also constructable.

File: externs/browser/html5.js [hunk 3]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+customElements.get = function(tagName) {};
+
+/**
+ * @param {string} tagName
+ * @return {Promise<!CustomElement>}
+ */
+customElements.whenDefined = function(tagName) {};
+
+/** @type {boolean} */
+Event.prototype.composed;
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Should these be near the rest of the `Event` declarations?

File: externs/browser/html5.js (hunk 1 above)
Comment (justinfagnani) [inline externs/browser/html5.js]: Could these be added as optional properties of `HTMLElement`?

File: externs/browser/html5.js [hunk 4]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+function HTMLSlotElement() {}
+
+/**
+ * @param {!{flatten: boolean}=} options
+ * @return {!Array<!Node>}
+ */
+HTMLSlotElement.prototype.assignedNodes = function(options) {};
+
+/** @type {?HTMLSlotElement} */
+Node.prototype.assignedSlot;
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Move with other `Node` declarations

File: externs/browser/html5.js [hunk 5]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+/**
+ * @return {!Array<!(Element|ShadowRoot|Document|Window)>}
+ */
+Event.prototype.composedPath = function() {};
+
+/**
+ * @param {!{mode: string}} options
+ * @return {!ShadowRoot}
+ */
+HTMLElement.prototype.attachShadow = function(options) {};
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Move with other `HTMLElement` declarations

File: externs/browser/html5.js [hunk 6]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+/** @type {boolean} */
+Event.prototype.composed;
+
+/**
+ * @return {!Array<!(Element|ShadowRoot|Document|Window)>}
+ */
+Event.prototype.composedPath = function() {};
+
+/**
+ * @param {!{mode: string}} options
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Is there a way to type `mode` as being either `'open'` or `'closed'`?

File: externs/browser/html5.js [hunk 7]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+ * @param {!{flatten: boolean}=} options
+ * @return {!Array<!Node>}
+ */
+HTMLSlotElement.prototype.assignedNodes = function(options) {};
+
+/** @type {?HTMLSlotElement} */
+Node.prototype.assignedSlot;
+
+/** @constructor */
+function InputDeviceCapabilities(){}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: are these last three related to the others?

File: externs/browser/html5.js [hunk 8]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+ * @constructor
+ * @extends {HTMLElement}
+ */
+function HTMLSlotElement() {}
+
+/**
+ * @param {!{flatten: boolean}=} options
+ * @return {!Array<!Node>}
+ */
+HTMLSlotElement.prototype.assignedNodes = function(options) {};
```
Comment (justinfagnani) [inline externs/browser/html5.js]: add `HTMLSlotElement.prototype.name`: https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/name

File: externs/browser/html5.js [hunk 9]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+
+/** @type {function()|undefined} */
+CustomElement.prototype.disconnectedCallback;
+
+/** @type {function(string, ?string, ?string, ?string)|undefined} */
+CustomElement.prototype.attributeChangedCallback;
+
+/**
+ * @see https://w3c.github.io/webcomponents/spec/custom/#custom-elements-api
+ * @const
```
Comment (justinfagnani) [inline externs/browser/html5.js]: I don't think it's actually const.

File: externs/browser/html5.js (hunk 1 above)
Comment (justinfagnani) [inline externs/browser/html5.js]: This is missing adoptedCallback

File: externs/browser/html5.js [hunk 10]
```diff
@@ -4286,3 +4288,102 @@ Plugin.prototype.length;
...
+/** @type {?HTMLSlotElement} */
+Node.prototype.assignedSlot;
+
+/** @constructor */
+function InputDeviceCapabilities(){}
+
+/** @type {boolean} */
+InputDeviceCapabilities.prototype.firesTouchEvents;
+
+/** @type {!InputDeviceCapabilities} */
```
Comment (azakus) [inline externs/browser/html5.js]: This is nullable

File: externs/browser/html5.js (hunk 7 above)
Comment (azakus) [inline externs/browser/html5.js]: These could probably be moved to an event extern.

File: externs/browser/html5.js (hunk 6 above)
Comment (ChadKillingsworth) [inline externs/browser/html5.js]: You can add an enum to "hint" that those are the only possible values, but it's not really enforced.

File: externs/browser/html5.js (hunk 2 above)
Comment (ChadKillingsworth) [inline externs/browser/html5.js]: If this is a constructible type (which provides better type checking anyway), than it doesn't need to be const. Const is used to tell the compiler that a symbol is a namespace.

File: externs/browser/html5.js (hunk 3 above)
Comment (ChadKillingsworth) [inline externs/browser/html5.js]: In an ideal world, yes. But the event specific externs are all tied to a specific spec version (like DOM level 3 events). The HTML5 externs have become the dumping grounds for living standard type definitions.

I'd like to change this (along with many other things), but I think it's beyond the scope of this PR.

File: externs/browser/html5.js (hunk 1 above)
Comment (ChadKillingsworth) [inline externs/browser/html5.js]: I really wanted to add these directly to HTMLElement - it significantly simplifies the type information. I just couldn't determine if that was correct from the spec and the Polymer project externs had them separated.

I've switched them to be properties on HTMLElement which completely removes the need for a `CustomElement` type and even a `CustomElementInterface`.

**Comment by ChadKillingsworth** [comment]
Ready for review again. In an ideal world, we would separate out a lot of these types (particularly node and events) into specific files. But html5.js is where all the living standard type updates have gone that aren't part of their own distinct spec (and even some that are).

**Comment by dimvar** [comment]
@ChadKillingsworth let me know when you address all comments, and I'll bring the PR in for testing.

**Comment by ChadKillingsworth** [comment]
@dimvar I think this is good to go.

File: externs/browser/html5.js [hunk 11]
```diff
@@ -1297,9 +1301,34 @@ SharedWorkerGlobalScope.prototype.name;
  */
 SharedWorkerGlobalScope.prototype.onconnect;
 
+/** @type {!Array<string>|undefined} */
+HTMLElement.observedAttributes;
+
+/**
+ * @param {!Document} oldDocument
+ * @param {!Document} newDocument
+ * @return {!ShadowRoot}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: adoptedCallback returns void

File: externs/browser/html5.js [hunk 12]
```diff
@@ -1297,9 +1301,34 @@ SharedWorkerGlobalScope.prototype.name;
...
+HTMLElement.prototype.adoptedCallback = function(oldDocument, newDocument) {};
+
+/**
+ * @param {!{mode: string}} options
+ * @return {!ShadowRoot}
+ */
+HTMLElement.prototype.attachShadow = function(options) {};
+
+/** @type {function(string, ?string, ?string, ?string)|undefined} */
+HTMLElement.prototype.attributeChangedCallback;
```
Comment (justinfagnani) [inline externs/browser/html5.js]: Can this have the arguments named for doc purposes, like adoptedCallback above?

File: externs/browser/html5.js [hunk 13]
```diff
@@ -4286,3 +4315,64 @@ Plugin.prototype.length;
 
 /** @type {string} */
 Plugin.prototype.name;
+
+/**
+ * @see https://w3c.github.io/webcomponents/spec/custom/#custom-elements-api
```
Comment (justinfagnani) [inline externs/browser/html5.js]: That's not the current spec. CE is folded directly into the HTML spec now: https://html.spec.whatwg.org/multipage/scripting.html#custom-elements

File: externs/browser/html5.js [hunk 14]
```diff
@@ -4286,3 +4315,64 @@ Plugin.prototype.length;
...
+ * @constructor
+ */
+function CustomElementsRegistry() {}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: The name is `CustomElementRegistry`. Singular `Element`. https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-api

File: externs/browser/html5.js [hunk 15]
```diff
@@ -4286,3 +4315,64 @@ Plugin.prototype.length;
...
+ * @param {string} tagName
+ * @param {!function(new:HTMLElement)} klass
+ * @param {{extends: string}=} options
+ * @return {!HTMLElement}
+ */
+CustomElementsRegistry.prototype.define = function (tagName, klass, options) {};
+
+/**
+ * @param {string} tagName
+ * @return {?HTMLElement}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: This returns the constructor, not an instance. Is that type annotation correct?

File: externs/browser/html5.js [hunk 16]
```diff
@@ -4286,3 +4315,64 @@ Plugin.prototype.length;
...
+ */
+CustomElementsRegistry.prototype.get = function(tagName) {};
+
+/**
+ * @param {string} tagName
+ * @return {Promise<!HTMLElement>}
```
Comment (justinfagnani) [inline externs/browser/html5.js]: same here, re: constructor type

**Comment by ChadKillingsworth** [comment]
@justinfagnani Nice catches. Updated.

**Comment by justinfagnani** [comment]
LGTM!

@azakus is going to try these in place of his externs to make sure everything is working. Let's wait on that before merging.

**Comment by dimvar** [comment]
ok, just ping this thread when the PR is ready to merge

**Comment by ChadKillingsworth** [comment]
Testing this on my projects revealed that `Element.prototype.name` exists and the `HTMLSlotElement.prototype.name` definition created warnings in NTI. I removed it.

File: externs/browser/html5.js [hunk 17]
```diff
@@ -4286,3 +4319,60 @@ Plugin.prototype.length;
...
+/** @type {boolean} */
+Event.prototype.composed;
+
+/**
+ * @return {!Array<!(Element|ShadowRoot|Document|Window)>}
+ */
+Event.prototype.composedPath = function() {};
+
+/** @constructor */
+function InputDeviceCapabilities(){}
```
Comment (azakus) [inline externs/browser/html5.js]: InputDeviceCapabilities has another boolean property `pointerMovementScrolls`.

InputDeviceCapabilities can be constructed with an optional bag of `{firesTouchEvents: boolean|undefined, pointerMovementScrolls: boolean|undefined}`

Comment (ChadKillingsworth) [inline externs/browser/html5.js]: @azakus Fixed - checked the latest version of the spec as well.

**Comment by justinfagnani** [comment]
@dimvar ready to go

**Comment by dimvar** [comment]
Tyler is on rotation this week, reassigned to him to bring in and test.

**Comment by MatrixFrog** [comment]
This is blocked on a small NTI bug but I expect we'll get it in within a day or two.
